rfcs/text/0002_encrypted_attributes.md

```diff
@@ -0,0 +1,146 @@
+type and the desired encrypted fields. The plugin would define its own
+wrapper around the client that would encrypt the registered attributes and
+remove any secret attributes on saved objects returned by the wrapped client.
+
+```JS
+server.plugins.encryptedAttributes.registerType('action', ['secretParams']);
+```
+
+If the wrapper gets a request to create, or update a registered type.
+
+It would use the following procedure.
+- extract and delete the registered encrypted attributes
+- use the resulting id, type, and attributes as additional authentication data
+  encrypt the extracted attributes.
+- add the encrypted attributes back to the saved object and pass it to the
+  wrapped client.
+
+The secret attribute would be scrubbed from all get, bulkGet, and find
+operations. Using the following procedure
+
+- Forward call to wrapped client
+- If no results or results blank return
+- For each result
+  - if the result type is a secret type
+    - delete the encrypted attribute field
+  - otherwise keep the results intact
+
+The wrapper will be injected for all saved object clients.
+
+## Benefits
+
+None of the registered types will expose their encrypted details. The saved
+objects with their unencrypted attributes could still be obtained and searched
+on. The wrapper will follow all the security and spaces filtering of saved
+objects so that only users with appropriate permissions will be able to obtain
+the scrubbed objects or _save_ objects with encrypted attributes.
+
+## Possible Problems
+
+Consumers of scrubbed objects may try to modify them causing the encrypted
+attributes to be dropped. But this could be prevented by disallowing updates
+from the default saved objects client. And instead force all updates to go
+through the plugin directly.
+
+## Decrypting attributes
+
+In order to get the resulting decrypted attributes back the plugin would require
+a method be registered which would be passed the saved object including the
+decrypted attributes then return an invokable method that would perform the
+retrieval of the saved object the decryption and invoking of the desired method.
+
+```JS
+const actionMethod = server.plugins.encryptedAttributes.buildDecryptionContext(id, type, (savedObject) => {
+  const { username, password } = savedObject.secretsParams;
+  ... // do integration with third party
+});
+... // later
+actionMethod(); // retrieve, decrypt, and invoke above method
+```
+
+## Benefits
+
+No explicit access to a method that takes in an encrypted string exists. And no
+access to the private key used is exposed through the plugin. If the type was not
+registered no decryption is possible. No need to handle the saved object with
+the encrypted attributes reducing the risk of accidentally returning it in a
+handler.
+
+# Drawbacks
+
+- This would add the ability to encrypt data on any saved object in the system.
+- Possibly add performance issues to retrieving saved objects.
```

Comment:
The encryption/decryption of the attributes should be the most CPU intensive operation, which should only occur when we're creating new objects which contain "secrets" or when we're decrypting the "secrets" in the background. We shouldn't be performing these operations for other saved object types, so this shouldn't impact these. 

Is there another performance implication which you were referring to which I'm not aware of?

Reply:
The biggest concern about performance I have is for the bulk operations. The wrapper essentially needs to iterate all returned objects regardless of type, generally this does not present a problem on small 10, 100, 1000 sized return sets. I was a little concerned about the >=10k results but doing some benchmarks it should not add significant time to be concerned with. More time will probably be spent waiting for results and parsing them than in this iteration. 

I don't think it will be a problem if the normal usage and calls are not returning that many hits which I don't think they are but I am not that sure.

Reply:
I don't think another O(n) operation is that problematic in this use-case.

Reply:
Well its not zero additional time and there is a lot more code at play than just a simple array iteration. For about 10k elements it'll add 2-3ms if doing 100k it'd be 20-30ms, I don't think its anything to stop the work just not as optimal as doing all the logic in a single iteration over the elements but we cannot really have that type of processing with the bulk requests and wrappers the way their structured. So it might not have been worth mentioning.